spec/message.ebnf

```diff
@@ -1,25 +1,20 @@
-Message ::= Declaration* ( Pattern | Selector Variant+ )
+Message ::= (s? Declaration)* s? Body s?
 
-/* Preamble */
-Declaration ::= 'let' WhiteSpace Variable '=' '{' Expression '}'
-Selector ::= 'match' ( '{' Expression '}' )+
+Declaration ::= 'let' s Variable s? '=' s? '{' s? Expression s? '}'
+Body ::= Pattern | Selector (s? Variant)+
 
-/* Variants and Patterns */
-Variant ::= 'when' ( WhiteSpace VariantKey )+ Pattern
-VariantKey ::= Literal | Nmtoken | '*'
-Pattern ::= '{' (Text | Placeholder)* '}' /* ws: explicit */
+Pattern ::= '{' (Text | Placeholder)* '}'
+Selector ::= 'match' (s? '{' s? Expression s? '}')+
+Variant ::= 'when' (s Key)+ s? Pattern
+Key ::= Nmtoken | Literal | '*'
 
-/* Placeholders */
-Placeholder ::= '{' (Expression | Markup | MarkupEnd)? '}'
+Placeholder ::= '{' s? (Expression | Markup | MarkupEnd)? s? '}'
```

Comment:
We can make the grammar be LL(2) without backtracking by forbidding empty placeholders:

```suggestion
Placeholder ::= '{' s? (Expression | Markup | MarkupEnd) s? '}'
```

I recall previous discussions with @eemeli about potentially allowing comment-only or metadata-only placeholders in the future. Any thoughts on this?

Reply:
My current thinking is that it'd be fine to forbid empty placeholders in the syntax. In the data model I'd probably allow for them, but that's really an entirely separate question.

On a separate topic on this same line: Should we include something like this, to make the mapping to a data model more direct?

```suggestion
Placeholder ::= '{' s? PlaceholderBody s? '}'
PlaceholderBody ::= Expression | Markup | MarkupEnd
```

Reply:
Ah, an interesting point about being strict in the syntax and lenient in the data model. I like the approach.

Reply:
Re. your other question: I don't think `PlaceholderBody` is needed. The grammar only describes a _parse tree_ (also known as a _concrete syntax tree_, CST). The CST will require further transformation before it becomes an AST. This is why I generally didn't attempt to map BNF productions to AST nodes.

Reply:
I filed #345 to discuss empty placeholders. Perhaps we can do it even before this PR is ready to merge.